Initial UB04 claims.

sjpadgett opened the pull request:
Stage 1 for institutional billing.

Up for grabs demo is here:
http://www.open-emr.org/wiki/index.php/Development_Demo#192.168.1.133

bradymiller commented on composer.json:
```diff
@@ -1,84 +1,84 @@
+		"symfony/yaml" : "2.8.*",
+		"phpoffice/phpexcel" : "1.8.1",
+		"ezyang/htmlpurifier" : "4.9.3",
+		"knplabs/knp-snappy" : "^0.5.0"
```
feel free to bring in the knp-snappy library as a separate PR to get it into the main codebase now

bradymiller commented on interface/billing/billing_report.php:
```diff
@@ -1,43 +1,28 @@
+ * @copyright Copyright (c) 2017 Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2016 Terry Hill <terry@lillysystems.com>
+ * @copyright Copyright (c) 2017 Brady Miller <brady.g.miller@gmail.com>
+ * @license https://www.gnu.org/licenses/agpl-3.0.en.html GNU Affero General Public License 3
```
Note the license above is the GPL, so should change this to:
``````
 * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3

sjpadgett replied:
Did not know we could change existing license. no problem.

sjpadgett replied:
I've heard of sleep walking but sleep coding! I actually don't remember changing. Sorry.

bradymiller commented on interface/billing/billing_report.php:
```diff
@@ -65,59 +49,60 @@
     html_header_show();
-} ?>
+}?>
+<script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-min-1-11-3/index.js"></script>
```
can we use jquery-min-3-1-1 here. If so, then would remove the following lines(the js/css libraries):
99-103 and 380-382
And then place the following line here:
``````
Header::setupHeader('datetime-picker')
``````
And then in the line below line 22 add the following:
``````
use OpenEMR\Core\Header;
``````

bradymiller replied:
Also would remove html_header_show if do this (lines 96-98)

bradymiller replied:
actually could also remove the common.js include at line 349 and then change the Header call to:
``````
Header::setupHeader(['datetime-picker','common']);
``````

sjpadgett replied:
Okay, able to do without any obvious problems. Will continue to test tho.

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,432 @@
+
+function get_payer_defaults($payerid)
+{
+    $p = sqlQuery("SELECT insurance_companies.* FROM insurance_companies WHERE insurance_companies.id = $payerid");
```
sql binding

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+ * @copyright Copyright (c) 2017 Jerry Padgett <sjpadgett@gmail.com>
+ * @license https://www.gnu.org/licenses/agpl-3.0.en.html GNU Affero General Public License 3
+ */
+/* $isAuthorized tells us if the form is for user UI or claim processing and provides another security check */
```
does this not need to require interface/globals.php?

bradymiller replied:
need to know above answer since using the Header class via autoloader needs globals.php.

bradymiller commented on interface/billing/ub04_helpers.php:
```diff
@@ -0,0 +1,77 @@
+        <tr>
+            <th>Provider</th>
+            <th>User Id</th>
+            <th>NPI</th>
```
translations

bradymiller commented on interface/billing/ub04_helpers.php:
```diff
@@ -0,0 +1,77 @@
+    $data = json_encode($row);
+?>
+<tr>
+    <td><button onclick='sendSelection(<?php echo $data;?>)'><?php echo xlt($row['fname'] . ' ' . $row['lname'])?></button></td>
```
don't translate name. use text().

bradymiller commented on interface/billing/ub04_helpers.php:
```diff
@@ -0,0 +1,77 @@
+<tr>
+    <td><button onclick='sendSelection(<?php echo $data;?>)'><?php echo xlt($row['fname'] . ' ' . $row['lname'])?></button></td>
+    <td><?php echo $row['id'] ?></td>
+    <td><?php echo $row['npi'] ?></td>
```
text() for above 2 lines

bradymiller commented on library/pdf_creator.class.php:
```diff
@@ -0,0 +1,86 @@
+        } else {
+            $bit = "64bit";
+        }
+        try {
```
will this support OSX?

sjpadgett replied:
yes but I pulled it because I could not (more likely, would not:)) test. I'll put back and make the pdf stuff a separate PR. I'm going to attempt using sub module to keep the binaries out of main git. I also really think sub module or fit will solve my nagging install issue with my node services. Really glad Robert brought the git management stuff up. Definitely on right track there.
Git is only my 100 and one source control I've had to learn over the years. I just want to code. Oh, and complain, of course :)

bradymiller replied:
Prob makes more sense to just bring in all the wkhtmltopdf binaries via composer(looks like that is what is suggested on their repo). Note that even if you create it as a submodule, all the binaries will still end up in the package release, which is why I am a bit unclear to the benefits of this. Whereas composer can be more flexible in this regard(that being said, would still plop all the binaries via composer into the vendor path).

sjpadgett replied:
Problem with using composer here is I can't find any trusted 1,2 or 3 places that have the binaries. h4cc has linux but is not current. We need somewhere to maintain our dependencies where we have control over them so, this is where I was hoping sub modules could help.
Take nodejs as example. You get what the author thinks you should have and you usually end up with a bunch of bloat. There is more to it than my poor explanation here but I think you get the gist.
I think we should create our own repositories for custom dependencies and composer could be used to install them.

bradymiller replied:
I looked at https://github.com/KnpLabs/snappy and I see what you mean. Looks like the main source is here: https://wkhtmltopdf.org/
Which is then built into binaries for a bunch of different os's in different places.
Quick question here; will the binary for example in ubuntu be different than other linux flavors?

This strikes as our very good first case to create a simple new github repo (wkhtmltopdf) which can then be loaded/updated into OpenEMR via composer(thus all work is done in this wkhtmltopdf upstream repo and then just brought into the OpenEMR codebase via composer). So will basically store all the possible binaries for this here (also then gives way for users to even build binaries for other os's and then contribute them to this custom wkhtmltopdf repo).

bradymiller replied:
@sjpadgett, Note that since your a member of OpenEMR group, you can create repos, so feel free to do this as needed for this work. Very exciting; openemr is possibly about to have its own custom composer module.

bradymiller commented:
very exciting!
Just did a quick review. 
Will do more detailed review/testing in the near future.

bradymiller commented:
Note @stephenwaite just brought in some code into the codebase that converted the Claim class to PSR2/4 and brought it into the autoloader. Should only be a couple minor adjustments needed in your code. Would rec rebasing your code on top of most recent codebase and then fixing the minor merge conflicts that will be in composer.json , library/Claim.class.php (now at library/billing/src/Claim.php ), and sql/5_0_0-to-5_0_1_upgrade.sql

bradymiller commented on library/gen_x12_837i.inc.php:
```diff
@@ -0,0 +1,1308 @@
+ * @author  Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2017 Jerry Padgett <sjpadgett@gmail.com>
+ * @license https://www.gnu.org/licenses/agpl-3.0.en.html GNU Affero General Public License 3
+ */
```
Place the following line (after you rebase to the most recent codebase) and ensure you have a blank line before and after this line:
``````
use OpenEMR\Billing\Claim;
``````

bradymiller commented on library/pdf_creator.class.php:
```diff
@@ -0,0 +1,86 @@
+<?php
+/**
+ * PdfCreator class
```
ok, lets turn this into a PSR2/PSR4 class that is autoloaded :)

bradymiller replied:
first, move this script to:
library/pdf/src/PdfCreator.php

bradymiller commented on library/pdf_creator.class.php:
```diff
@@ -0,0 +1,86 @@
+ * @author  Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2017 Jerry Padgett <sjpadgett@gmail.com>
+ * @license https://www.gnu.org/licenses/agpl-3.0.en.html GNU Affero General Public License 3
+ */
```
Place the following (needs to be above the use command and it needs an empty line both above and below it):
``````
namespace OpenEMR\Pdf;
``````

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,432 @@
+require_once("../globals.php");
+require_once("$srcdir/Claim.class.php");
+require_once("$srcdir/pdf_creator.class.php");
+include_once("$srcdir/gen_x12_837i.inc.php");
```
place the following here(part of the PdfCreator class PSR2/PSR4 conversion):
``````
use OpenEMR\Pdf\PdfCreator;
``````

bradymiller replied:
also change above include_once to require_once

bradymiller replied:
also will need this here:
``````
use OpenEMR\Billing\Claim;
``````

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,432 @@
+ */
+require_once("../globals.php");
+require_once("$srcdir/Claim.class.php");
+require_once("$srcdir/pdf_creator.class.php");
```
remove above 2 lines (after rebase to most recent master and converting PdfCreator to PSR2/PSR4) since these are now autoloaded classes :)

bradymiller replied:
Also. Almost forgot. Since removing the Claim.class.php include, need to add the following:
``````
require_once("$srcdir/invoice_summary.inc.php");
``````

bradymiller commented on composer.json:
```diff
@@ -1,84 +1,84 @@
+			"OpenEMR\\Menu\\" : "library/menu/src",
+			"OpenEMR\\Calendar\\" : "library/calendar/src",
+			"OpenEMR\\Sample\\" : "library/sample/src",
+			"OpenEMR\\Reminder\\" : "library/reminder/src"
```
place a comma above and then add the following line(for PdfCreator PSR2/PSR4 class upgrade):
``````
"OpenEMR\\Pdf\\" : "library/pdf/src"
``````
Then when you do above and the other items to convert the PdfCreator class to PSR2/PSR4, then do the following composer command to wire in your new class into the autoloader:
``````
composer dump-autoload -o
``````

sjpadgett replied:
Okay thanks (you are really good at this don't ch'ya know:)). Let me tiny up this commit, take an aspirin and pull the PDF stuff for a separate commit.

bradymiller replied:
btw, sorry for working on this stuff at the same time as you. The initial goal was to do some focused bug fixes in X12, however, we then got carried away and ended up converting Claim class to PSR2/4 :)

sjpadgett replied:
Really cant be helped I guess. I just need to be a little faster:)
Why is it important to make the previous claims public functions cleanDate and x12Clean, private?
These are handy outside the class.

bradymiller replied:
When make a class PSR4 and autoload, best to have no baggage in the class file (ie. requires or function libraries). You can definitely add those functions in one of the included libraries if you need to use them outside the Claim class.

bradymiller replied:
Another option if always including the claim class when need those functions is to convert those 2 classes to (static if needed) public functions.

sjpadgett replied:
I accept that but why not public instead? why dup function elsewhere?

bradymiller replied:
At the time, since it was only in the Claim class script, I may of incorrectly assumed that it wasn't used elsewhere. Definitely feel free to make public (and static if needed) to reduce dup code.

sjpadgett replied:
I think I'll do that because there are institutional items that really need to go in that class but I will deal with that when I get to serious X12 testing.

bradymiller commented:
Also, just let me know if need help with the rebase to most recent master. After you do this, then when you push code to this PR will need to use --force and command will look like this:
``````
git push --force origin institutional-billing
``````

sjpadgett commented:
I think I'm okay. I fetched, pulled and rebased into branch. Pushed master to origin, checked out branch, cleaned up conflicts made mods and am testing. I am however going to still included the snappy libs on my next commit to keep things sane.

sjpadgett commented:
@bradymiller Does the Header class not auto bootstrap. Seems I have to include bootstrap in the billing_report.

bradymiller commented:
yes, the Header brings in bootstrap automaticall (also supports rtl when needed). For sanity checking, rec viewing source from browser to ensure the libraries are there. A common error is not placing the list of items within the Header call within an array.

sjpadgett commented:
@bradymiller No idea where I am any longer. Hope the commit is not too much hassle to review and I'm sure I've missed something but I'm off to figure out composer stuff for pdf.

sjpadgett commented:
Also will add osx to pdf class when I do the pdf binaries PR.

bradymiller commented:
Very close, but it looks like you took a relatively painful route :)

I noted a couple places where this will break. Let me know if it's ok for me to tidy this up a bit, get it into 1 commit, and set up composer for your new class.

bradymiller commented:
also happy to bring the knp-snappy into the main codebase asap. just give me the word.
(then you can focus yourself on the prize, which is ub04 :) )

sjpadgett commented:
Painful is the only git I know! Go ahead, if you will, and get me straight.

bradymiller commented:
Did 2 things:
1. Just committed in the knp-snappy stuff into main codebase via composer
2. Cleaned up your code and rebased into 1 commit on top of most recent codebase here:
https://github.com/bradymiller/openemr/commits/institutional-billing

To replace your PR code with this new branch, do the following:
``````
git checkout institutional-billing
git checkout -b backup-institutional-billing
git branch -D institutional-billing
git remote add bradymiller https://github.com/bradymiller/openemr.git
git fetch bradymiller
git checkout bradymiller/institutional-billing
git checkout -b institutional-billing
git push --force origin institutional-billing
``````

sjpadgett commented:
Think of a programming language and I've written code using it but, for some reason, when it comes to Git, my brain says Enough! No ambition to learn it. Anyway thanks. Still don't know what I did so horrible. Maybe went wrong on rebase conflicts.

stephenwaite commented on interface/billing/ub04_helpers.php:
```diff
@@ -0,0 +1,77 @@
+    $data = json_encode($row);
+?>
+<tr>
+    <td><button onclick='sendSelection(<?php echo $data;?>)'><?php echo text$row['fname'] . ' ' . $row['lname'])?></button></td>
```
hi @sjpadgett , enjoying your progress with @bradymiller 's guidance

you probably saw this but travis didn't like your text function without an opening paren

cheers

sjpadgett replied:
I don't get it. I run this stuff through ci before I submit and still get issues. I don't trust if I miss typed or beautifier did it. Anyway, thanks for looking out.

bradymiller replied:
note your code passes the styling stuff. this issue with travis is part of the check for php errors (php parsing error).

bradymiller commented:
hi @sjpadgett ,
Rebasing is probably the toughest thing to master in git. Your git skills are already strong (I suggest looking at your old git work a year ago and you will see how far you have come). Soon, you'll master git rebase. A strategy to use for git rebase is to:
1) update master branch to most recent code
2) on your working branch, then rebase all your work into 1 commit
3) on your working branch, then do `git rebase master`
(then use `git status` to help work through the merge conflicts)

sjpadgett commented:
Prob went wrong when I fetched and pulled upstream from working master then rebased into branch pushed to origin/master then checked out branch. Last time I tried rebase inside branch things didn't work out so well.
I get frustrated that I spend loads of time testing to just introduce bugs on commit. 
No worries though as my computer is still on my desk and not in the toilet yet. So if you see blurp blurp next commit, you know whats up:)

sjpadgett commented:
@bradymiller I was able to prove the concept of using composer to handle pdf binaries using a bitbucket repo on my account so I created an openemr repo and will clone into it.  Should I go ahead and open a PR for us to work from or do you want to work from the new openemr repo?

bradymiller commented:
Ok with either and do whatever is easiest for you at this point to keep your project moving. We can adjust things(such as composer repo name/path) as needed without affecting your workflow in the ub04 project. Guessing both @robertdown and @TheToolbox will provide guidance on how to best optimize the new composer repo and will function as a good model for others to build openemr custom composer modules/packages.

bradymiller commented:
just noted the wkhtmltopdf-openemr repo you created. That's a fine place to put it. Rec just putting it there and then putting it into your PR via composer. And then we (I'm glad to spend time focusing on this while you work on ub04) can work on getting it to be consistent with composer best standards.

sjpadgett commented:
@bradymiller I believe the composer is working well and was easier than I expected. I also like the repo solution and probably going forward sub modules become mute. A believe a more precise composer install may be possible(sense os environment and decide required)  but I'm to new to composer and will need to research. Also I will do a fresh install from this branch on my linux server to see where I stand (feels like some items may be missing from commits). Thanks for your help and it is always a pleasure working with you.

sjpadgett commented:
@stephenwaite I'm getting ready to go through my X12 based on your recent work in claims, thanks for that. We don't do electronic billing so I don't have a test bed. Do you know if there are sites that allow test claims?

stephenwaite commented:
hi @sjpadgett it's usually through a clearinghouse or payer that you already have an established relationship

would you like me to reach out to mine and then we could work together on submitting and fixing the x12 errors?

sjpadgett commented:
@bradymiller Thought so. Hate to bother you but it would be great to know where I stand. Let me review my 837I, implement some of Stephens cleanup and setup a test encounter on my server and I will get back to you. Thanks.

stephenwaite commented:
like @bradymiller said, be glad to help with anything that keeps you cruising :)

sjpadgett commented:
@stephenwaite I think I found an error in your gen_x12_837 cleanup at line 1053 where
if ($ndc) {
        ++$edicount;
    }
appears to be closed to soon. It is advancing the count but what the conditional is for (ndc) is outside the conditional. i.e.
   $out .= "LIN" . // Drug Identification. Page 500+ (Addendum pg 71).
    "*" .         // Per addendum, LIN01 is not used.
    "*" . "N4" .
    "*" . $ndc .
    "~\n";

    if (!preg_match('/^\d\d\d\d\d-\d\d\d\d-\d\d$/', $ndc, $tmp) && !preg_match('/^\d{11}$/', $ndc)) {
        $log .= "*** NDC code '$ndc' has invalid format!\n";
    }

    ++$edicount;
    $out .= "CTP" . // Drug Pricing. Page 500+ (Addendum pg 74).
    "*" .
    "*" .
    "*" .
    "*" . $claim->cptNDCQuantity($prockey) .
    "*" . $claim->cptNDCUOM($prockey) .
    // Note: 5010 documents "ME" (Milligrams) as an additional unit of measure.
    "~\n";
Should I go ahead and fix it?

stephenwaite commented:
sure @sjpadgett , nice catch, thanks

 are you going to create a gen_x12_837**I** script or work your magic inside this script?

sjpadgett commented:
@stephenwaite Also at line 1046 the procedure loop has been closed before NDC (recorded in fee sheet with HCPCS code) has been included. Looks to me that possibly only one code would ever get transmitted and rendering provider is include with each line item previously.

I have my own x12 but I'm going through both to sort out the 5010 stuff like you did.

sjpadgett commented:
@stephenwaite I believe at line 1130:
$out .= "REF" .
            "*" . $claim->providerNumberType($prockey) .
            // Note: 5010 documents that type 1D (Medicaid) is changed to G2.
            "*" . $claim->providerNumber($prockey) .
            "~\n";
is not required for 5010 according to previous logic. I would recommend going back to original commit and double check diff where CMS_5010 has been edited out.

sjpadgett commented:
@stephenwaite Guess who? ;) Okay I fixed these but double check me when I commit with my X12. I'll do separate commit for each.

stephenwaite commented:
okay, nice, thanks for the cleanup @sjpadgett

sjpadgett commented:
@stephenwaite I'm a little concerned that much of the ECLAIMS traps have been removed. Was eclaims a 4010 issue or no longer supported for cms-1500?

stephenwaite commented:
hi @sjpadgett those were 4010 eclaims specs and not sure they're even in business anymore

bradymiller commented on composer.json:
```diff
@@ -1,86 +1,92 @@
+			"doctrine/couchdb" : "dist"
+		}
+	}
+}
```
There is something odd going on with the formatting of this script. Did this get converted to tabs? If so, rec changing to spacing(guessing you editor is automatically changing your json scripts to use tabs maybe).

sjpadgett replied:
Could be, I just turned that feature on in my ide. I'll check and resubmit after you finish on other commits. Seems like every editor I have I must specifically change to spaces.

sjpadgett replied:
Turned on tab to space in my json editor just to have the composer tool over write so, I didn't need the darn tool anyway! Matter of fact, notepad++ is looking good again!

bradymiller commented on interface/billing/billing_report.php:
```diff
@@ -65,59 +52,55 @@
-<script>
+<?php Header::setupHeader(['datetime-picker','common']); ?>
+<link rel="stylesheet" href="<?php echo $GLOBALS['assets_static_relative'] ?>/bootstrap-3-3-4/dist/css/bootstrap.min.css">
+<script src="<?php echo $GLOBALS['assets_static_relative']; ?>/bootstrap-3-3-4/dist/js/bootstrap.min.js" type="text/javascript"></script>
```
rec trying to remove above 2 lines again. i confirmed that these two lines are already included in the Header::setupHeader call above (in browser on the page, did a Frame->View Source).

bradymiller commented on interface/billing/billing_report.php:
```diff
@@ -347,303 +340,336 @@ function MarkAsCleared(Type)
+        <div class='title'><?php echo xlt('Billing Manager') ?></div>
+    </p>
+    <form name='the_form' method='post' action='billing_report.php' onsubmit='return top.restoreSession()' style="display: inline">
+        <?php Header::setupHeader('datetime-picker'); ?>
```
can remove this line. you already did this above.

sjpadgett replied:
That's the way I originally had it and had trouble with bootstrap losing scope. But, worked for me today. Maybe problem came from all my tests with composer.

bradymiller commented on interface/billing/billing_report.php:
```diff
@@ -705,14 +733,14 @@ class='link_submit'  ><?php echo '['.xlt('End Of Day Report').']' ?></a>
-            "pid=? AND " .
-            "activity = 1", array($iter['enc_encounter'],$iter['enc_pid']));
+        if (! $iter['id']) {
+            $res = sqlQuery("SELECT count(*) AS count FROM billing WHERE " . "encounter = ? AND " . "pid=? AND " . "activity = 1", array(
```
there's appears to be some weirdness (ie. auto changes) where lines, such as above are getting combined. there are some more obvious examples of this below in this script.

bradymiller replied:
This is definitely an odd thing that has happened. It seems like there was auto changed done here and below that concatenated these queries which has caused them from going from readable to no longer readable. Rec. looking at the script with the ignore whitespacing via this link to see what I mean:
https://github.com/openemr/openemr/pull/1014/files?w=1

bradymiller commented on interface/billing/billing_report.php:
```diff
@@ -705,14 +733,14 @@ class='link_submit'  ><?php echo '['.xlt('End Of Day Report').']' ?></a>
-            "pid=? AND " .
-            "activity = 1", array($iter['enc_encounter'],$iter['enc_pid']));
+        if (! $iter['id']) {
+            $res = sqlQuery("SELECT count(*) AS count FROM billing WHERE " . "encounter = ? AND " . "pid=? AND " . "activity = 1", array(
```
for a more extreme example, check out line 850

sjpadgett replied:
I honestly don't know Brady. I know I did it but every since I upgraded my eclipse to ensure I have latest PSR stuff, I get weird stuff. Please don't tell on me to the other developers.:)

I'll try to sort it out.

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,433 @@
+require_once("$srcdir/invoice_summary.inc.php");
+
+use OpenEMR\Billing\Claim;
+
```
remove the empty line between these 2 use commands

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,433 @@
+        echo $ub04id;
+        exit();
+    }
+    die(xl('Do not know what to do!'));
```
xlt

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,433 @@
+        $ub04id = json_encode($ub04id);
+        $isAuthorized = true;
+        ob_start();
+        include(dirname(__file__) . "/ub04_form.php");
```
best to use `require` if this is needed

bradymiller commented on interface/billing/ub04_dispose.php:
```diff
@@ -0,0 +1,433 @@
+
+    $isAuthorized = true;
+    ob_start();
+    include(dirname(__file__) . "/ub04_form.php");
```
best to use `require` if this is needed

sjpadgett replied:
Sometimes we don't want to stop a script on a dependency that will show itself in a more peaceful manner later. Still I changed.

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+<meta http-equiv="X-UA-Compatible" content="IE=Edge" />
+<meta charset="utf-8" />
+<?php if ($isAuthorized !== true) {?>
+<script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-min-1-11-3/index.js"></script>
```
Will jquery-min-3-1-1 work here?

sjpadgett replied:
Hung browser 'til we had a talk.

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+pid = <?php echo $pid;?>;
+encounter = <?php echo $encounter;?>;
+isTemplate = <?php echo ($isAuthorized === true ? $isAuthorized : 0); ?>;
+ub04id = <?php echo $ub04id;?>;
```
use attr() to escape 5 echo'd variables above

sjpadgett replied:
Escaping ub04id(a json encoded array of objects) will blow up the array. So a no there. Will escape the others but I never see the sense escaping variables that really don't require it. A boolean for example or integers.

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+        if (xhr.status === 200) {
+            console.log(this.responseText?this.responseText:'no response');
+            if(this.responseText == 'done')
+                alert("<?php echo xl("Save Completed")?>");
```
use xls()

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+#ub04id426{left:436px; top:1155px; width:132px; height:14px; color: rgb(0,0,0); text-align:right;}
+#ub04id427{left:594px; top:1155px; width:170px; height:14px; color: rgb(0,0,0); text-align:left;}
+#ub04id428{left:790px; top:1155px; width:129px; height:14px; color: rgb(0,0,0); text-align:left;}
+#ub04id429{left:19px;  top:1155px; width:263px; height:14px; color: rgb(0,0,0); text-align:left;}
```
I am afraid to ask how much time it took for you to come up with all the px measurements :)

sjpadgett replied:
Write code that writes code and math is our friend!

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+<body onload="adjustForm();">
+<div class="container" id="formContainer">
+<?php if ($isAuthorized !== true) {?>
+<h3 class='formhide'><em>Claim Edit </em><button class="btn btn-xs btn-warning" onclick="myZoom()" >Zoom</button></h3>
```
translate

sjpadgett replied:
I'm afraid to:)

bradymiller replied:
face your fears :)

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+<?php if ($pid && $encounter) {?>
+    <button class="btn btn-xs btn-success" onclick="disposeSave('form')" title="Save for printing with form">Pdf With Form</button>
+    <button class="btn btn-xs btn-success" onclick="disposeSave('noform')" title="Save for printing to a pre printed sheet">Pdf Without Form</button>
+    <button class="btn btn-xs btn-success" onclick="postClaim('batch_save')" title="Save claim for batch processing">Save Claim</button>
```
translate above

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+    <button class="btn btn-xs btn-success" onclick="disposeSave('noform')" title="Save for printing to a pre printed sheet">Pdf Without Form</button>
+    <button class="btn btn-xs btn-success" onclick="postClaim('batch_save')" title="Save claim for batch processing">Save Claim</button>
+    <?php } else {?>
+    <button class="btn btn-xs btn-success" onclick="postClaim('payer_save')">Save Payer</button>
```
translate

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+    <?php } else {?>
+    <button class="btn btn-xs btn-success" onclick="postClaim('payer_save')">Save Payer</button>
+    <?php } ?>
+    <button class="btn btn-xs btn-danger" onclick="resetClaim()" title="Reset claim form to Fee Sheet Version">Reset Version</button>
```
translate

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+    <?php } ?>
+    <button class="btn btn-xs btn-danger" onclick="resetClaim()" title="Reset claim form to Fee Sheet Version">Reset Version</button>
+    <button class="btn btn-info btn-xs" type="button" onclick="window.scrollTo(0, 0);$('#formhelp').toggle()"><?php echo xlt('Help'); ?></button>
+    <button class="btn btn-xs btn-danger" onclick="cleanUp()">Return</button>
```
translate

bradymiller commented on interface/billing/ub04_form.php:
```diff
@@ -0,0 +1,1404 @@
+<input type="text" data-dp   maxlength="10"  id="ub04id25" value=""  title="12. PATIENT ADMISSION/START OF CARE DATE (MMDDYY)"  name="ub04Block_24"/>
+<input type="text"   maxlength="2"  id="ub04id26" value=""  title="13. PATIENT ADMISSION HOUR"  name="ub04Block_25"/>
+<input type="text"   maxlength="1"  id="ub04id27" value=""  title="14. PATIENT PRIORITY (TYPE) OF VISIT"  name="ub04Block_26"/>
+<input type="text"   maxlength="1"  id="ub04id28" value=""  title="15. PATIENT POINT OF ORIGIN FOR ADMISSION OR VISI
```
agree with not translating these titles

bradymiller commented on interface/billing/ub04_helpers.php:
```diff
@@ -0,0 +1,77 @@
+}
+
+// Falling through for user dialog.
+$users = sqlStatementNoLog("SELECT id,fname,lname,npi FROM users WHERE " . "authorized=? AND active=?", array(1,1));
```
another place where eclipse likely combined 2 lines.

bradymiller commented on interface/billing/ub04_helpers.php:
```diff
@@ -0,0 +1,77 @@
+        <tr>
+            <th><?php echo xl('Provider')?></th>
+                <th><?php echo xl('User Id') ?></th>
+                <th><?php echo xl('NPI') ?></th>
```
use xlt in above 3 lines

bradymiller commented on interface/patient_file/encounter/superbill_custom_full.php:
```diff
@@ -260,12 +269,41 @@ function bucks($amount)
+    <?php if ($institutional) { ?>
+    <script type="text/JavaScript" src="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-min-3-1-1/index.js"></script>
+    <link href="<?php echo $GLOBALS['assets_static_relative']; ?>/jquery-ui-1-12-1/themes/base/jquery-ui.min.css" rel="stylesheet" type="text/css" />
+    <script type="text/javascript" src="<?php echo $GLOBALS['assets_static_relative'] ?>/jquery-ui-1-12-1/jquery-ui.min.js"></script>
```
to aovid confusion down the road, rec placing the above 3 asset with the other assets prior to the if $institutional

bradymiller commented on library/gen_x12_837.inc.php:
```diff
@@ -1043,95 +1043,98 @@ function gen_x12_837($pid, $encounter, &$log, $encounter_claim = false)
+            // REF*1C is required here for the Medicare provider number if NPI was
+            // specified in NM109.  Not sure if other payers require anything here.
+
+            // @TODO below appears to not belong in 5010??.
```
@stephenwaite , any thoughts on this? (btw, more clear if look at this with ignoring of whitespace changes: https://github.com/openemr/openemr/pull/1014/files?w=1 )

stephenwaite replied:
from my experience only medicaid requres it so could be recoded to 
`if ($claim->providerNumberType($prockey) == "G2") {
    ++$edicount;
    $out .= "REF" .
            "*" . $claim->providerNumberType($prockey) .
            "*" . $claim->providerNumber($prockey) .
            "~\n";
}`

sjpadgett replied:
I didn't format this but php phpcs.phar  -n --standard=ci/phpcs.xml showed errors so I ran it through: php phpcbf.phar  -n --standard=ci/phpcs.xml thus what we got.

stephenwaite replied:
the 3 comments and blank line in @bradymiller post can be removed
the comments above that are helpful

sjpadgett replied:
@stephenwaite Sorry, unclear what you mean. I did change medicaid require, it's the post can be removed the comments, I don't understand.

bradymiller commented on library/gen_x12_837i.inc.php:
```diff
@@ -0,0 +1,1244 @@
+ * @link    http://www.open-emr.org
+ * @author  Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2017 Jerry Padgett <sjpadgett@gmail.com>
+ * @license https://www.gnu.org/licenses/agpl-3.0.en.html GNU Affero General Public License 3
```
is this code was derived from the other gen_x12, then should use the GPL3 license instead:
``````
* @license https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3

bradymiller commented on library/globals.inc.php:
```diff
@@ -867,6 +867,27 @@ function gblTimeZones()
+            'num', // data type
+            '11', // default
+            xl('This is the default left print margin for UB04. It will adjust the final printed output left or right.')
+        ),
```
Do you think these should go in the Billing tab?
(I am not sure since billing items seem to be all over the place)

sjpadgett replied:
I just put them where the cms-1500 were. I can move both to billing but keeping where the feature is turned on may be best. Up to you.

bradymiller commented on sql/5_0_0-to-5_0_1_upgrade.sql:
```diff
@@ -510,4 +510,17 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
+#EndIf
+
+#IfMissingColumn claims submitted_claim
+ALTER TABLE `claims` ADD COLUMN `submitted_claim` TEXT NULL DEFAULT NULL COMMENT 'This claims form claim data';
```
remove `NULL DEFAULT NULL`
(TEXT should not be set to anything)

bradymiller commented on sql/5_0_0-to-5_0_1_upgrade.sql:
```diff
@@ -510,4 +510,17 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
+#EndIf
+
+#IfMissingColumn codes revenue_code
+ALTER TABLE `codes` ADD COLUMN `revenue_code` varchar(255) NOT NULL DEFAULT "" COMMENT 'Item revenue code';
```
quick likely dumb question.
why is this revenue_code varchar(255) while he one above is varchar(6)?

sjpadgett replied:
Because in codes I had plans to also write the revenue code description with the code. I'm not sure if that is still not a possibility. May as well keep at 6, Just keeping option open.

sjpadgett commented:
Are we close to being done beating up poor ole Jerry? :)

bradymiller commented on sql/database.sql:
```diff
@@ -312,6 +313,7 @@ CREATE TABLE `claims` (
   `process_file` varchar(255) default NULL,
   `target` varchar(30) default NULL,
   `x12_partner_id` int(11) NOT NULL default '0',
+  `submitted_claim` text default NULL,
```
remove `DEFAULT NULL`
(TEXT should not be set to anything)

bradymiller commented on sql/database.sql:
```diff
@@ -719,6 +721,7 @@ CREATE TABLE `codes` (
+  `revenue_code` varchar(255) NOT NULL default '',
   PRIMARY KEY  (`id`),
   KEY `code` (`code`),
   KEY `code_type` (`code_type`)
```
Note that these entries do not have COMMENTS while the entries int he upgrade script do.

bradymiller commented:
@sjpadgett ,
just completed a more detailed code review. looking good. will do more extensive testing this evening (I can't test the billing, but can test gui/buttons etc.).

sjpadgett commented:
okay, i'll add it.

On Tue, Aug 22, 2017 at 4:42 PM, stephen waite <notifications@github.com>
wrote:

> *@stephenwaite* commented on this pull request.
> ------------------------------
>
> In library/gen_x12_837.inc.php
> <https://github.com/openemr/openemr/pull/1014#discussion_r134596210>:
>
> > +            // Segment NM1,N3,N4 (Service Facility Location) omitted.
> +            // Segment REF (Service Facility Location Secondary Identification) omitted.
> +            // Segment NM1 (Supervising Provider Name) omitted.
> +            // Segment REF (Supervising Provider Secondary Identification) omitted.
> +            // Segment NM1,N3,N4 (Ordering Provider) omitted.
> +            // Segment REF (Ordering Provider Secondary Identification) omitted.
> +            // Segment PER (Ordering Provider Contact Information) omitted.
> +            // Segment NM1 (Referring Provider Name) omitted.
> +            // Segment REF (Referring Provider Secondary Identification) omitted.
> +            // Segments NM1*PW, N3, N4 (Ambulance Pick-Up Location) omitted.
> +            // Segments NM1*45, N3, N4 (Ambulance Drop-Off Location) omitted.
> +
> +            // REF*1C is required here for the Medicare provider number if NPI was
> +            // specified in NM109.  Not sure if other payers require anything here.
> +
> +            // @TODO below appears to not belong in 5010??.
>
> from my experience only medicaid requres it so could be recoded to
> if ($claim->providerNumberType($prockey) == "G2") { ++$edicount; $out .=
> "REF" . "*" . $claim->providerNumberType($prockey) . "*" .
> $claim->providerNumber($prockey) . "~\n"; }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openemr/openemr/pull/1014#discussion_r134596210>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxhkWX3LwFaEwApGMVudn8rJ6RUM42rks5saz1EgaJpZM4O505O>
> .
>



-- 
Jerry Padgett
Padgett Systems and Consulting
Brandon, Florida 33511
sjpadgett@gmail.com <sjpadgett@gmailo.com>
(239) 895-4333

sjpadgett commented:
@bradymiller While we're here do you want me to go ahead and consolidate the billing globals setup into billing tab?

stephenwaite commented:
hi @sjpadgett i'll be set to submit 837I test files through my clearinghouse tomorrow

sjpadgett commented:
@bradymiller Terrific, I'm working now on a sample for an outpatient surgery and a professional claim.  Once done you can go to my server and gen the download/or check me on the encounter. Thanks, this is going to help much.

sjpadgett commented:
@bradymiller I completely forgot I have a lookup table for all the various codes. Currently we're looking at around 1600 entries. So it is import the table or create an array for it. What's the preference?

stephenwaite commented:
@sjpadgett ok, will login create an x12 partner with my submitter credentials, is the site here
https://community.open-emr.org/t/proposed-ub04-cms-1450-claims/8747/2
?

sjpadgett commented:
Yes, I still need to put latest commits up but you should be good to go.

stephenwaite commented:
ok thanks @sjpadgett , it's me, stephen, that will be testing the 837I and will also be good to test the 837P with all of the work that's going on.  I can start the 837P today it's the I that the clearinghouse had to enable.  I'm currently configuring the x12 partner.

sjpadgett commented:
@stephenwaite Perfect. Do what ever you need on the server. Insurance co. are made up and Jason Binder is patient I kill mostly. Server does not reset so anything you do, persist.
Let me know any problem or mods you need and I can put them up on this P.R and thanks for your help, I appreciate it.

bradymiller commented:
Up for grabs demo for this PR is here:
http://www.open-emr.org/wiki/index.php/Development_Demo#192.168.1.133

bradymiller commented:
hi @sjpadgett ,
yep, would migrate that grouping of billing globals to the top of the Billing globals screen.
thanks, -brady

bradymiller commented:
hi @sjpadgett ,
Regarding the codes, is it 1 code set or a bunch of different code sets?
-brady

sjpadgett commented:
@bradymiller They are a consolidation of revenue, condition, occurrence, value and patient status codes. About 900 total. I went ahead and put them in an array and am writing the array search now. Figure wait 'till institutional becomes more mature in OpenEMR to distribute through out our code U.Is.
CMS has downloads of many HCPCS and revenue codes they maintain and release that we will need to eventually write import for. For now I'll just keep out of database and go with static include, if you agree.
I also added another save and stay button to the fee sheet setting up for an autosave. Fee sheet has been touched by so many it is somewhat of a task to figure out what really goes on there:) All in all though, it is 100 times better than 2 yrs ago.
I must say after going through latest master on my server, this release is looking very nice. Kudos to @robertdown and the others working the U.I for how clean and modern the application is shaping up.

Seems I'm awful chatty this morning, better go. New commit by this afternoon.

sjpadgett commented:
@bradymiller Can translate translate a variable from say lookup or do I have to translate the static array?
i.e. find 'term" $r = term; return xla($r);

bradymiller commented:
@sjpadgett , better to translate the title in the array directly so it's picked up in the translation pipeline and to keep the code clear. If this is regarding your code sets that you plan to add, would rec not even translating those for now.

stephenwaite commented:
hi @sjpadgett , just for the purposes of my testing will need to replace these lines
https://github.com/sjpadgett/openemr/blob/7b781406fa8fdb31e3a314ac947d51a032101746/library/gen_x12_837.inc.php#L85-L124
with this 
`$out .= "NM1*41*2*CARE MANAGEMENT SOLUTIONS*****46*030353360~\n" .
 $out .= "PER*IC*STEPHENWAITE*TE*8003718685*FX*8027705175*EM*STEPHEN.WAITE@CMSVT.COM~\n"`

sjpadgett commented:
@stephenwaite  I took patched out thru L125 and terminated new string ; .  I put up on server. If you want just e-mail me files of any changes and I can have on server in a jiff.

stephenwaite commented:
cool @sjpadgett , testing 837Ps, will keep you posted

sjpadgett commented:
@bradymiller Static ub codes and fee sheet enhancement pushed.

sjpadgett commented:
@stephenwaite I patched the 837I with same account info from 837P and created a X12 claim on Jason Binder 5/14 for Home health physical therapy X12 837I(last queued). I also created you a new claim for an extended office visit with a A1c lab for Jason Binder 8/28 30-50 for 837P.
I reviewed both in edi viewer and they look okay. Did you have specific file level errors? I did a little cleanup for insurance for Binder so maybe that will help!! Also, there may have been some procedures in 5/14 claim you submitted that may not have been suitable for 837P, don't know.
Let me know if you need anything.

sjpadgett commented:
@bradymiller Do you know who did the checksum in fee sheet? Is it new to 5.0.1?

sunsetsystems commented:
The checksum code is mine. Originally done to interface/forms/fee_sheet/new.php on 2013-12-10, and later moved to the FeeSheet class.  Questions?

sjpadgett commented:
Yes, thanks. I assume is was meant to prevent competing users record contention. Was this an issue in past or a prevention measure? Not looking to change but understand history some as I may need to copy and tailor to a specific Institutional fee sheet.

sunsetsystems commented:
Yes, to prevent conflicting updates of a fee sheet instance by different users at the same time. And yes motivated by past instances of such. The idea is to reject an update at save time if the fee sheet data at that time differs from what it was when the form was presented.

sjpadgett commented:
Okay. Don't know if a new issue(maybe I created) or one that's not being reported but I'm seeing issue even on new fee sheet but, over time, so much has been done to the sheet that adding a different animal like Institutional is asking for it. Anyway I'll figure out current issue. Thanks.

sjpadgett commented:
@bradymiller No idea why sql/5_0_0-to-5_0_1_upgrade.sql is conflicted. My source and commit code is okay. May have been... shoot who knows!
Anyway, I don't plan any more major additions. As far as 837I, all I can say is:
 if( ! $devgaveitashot ){exit()} elseif($devisconfused){ echo xla('Exactly how many different way's can a claim be submitted, Y'all?')};
So the integration with billing seems to be okay with a sensible work flow and some minor encroachment into Institutional billing. I would like to set up a separate fee sheet for institution because it follows a very different path than CMS1500 but, need to see what happens.
UB04 paper submissions should have no problems due to in form editing and printing exactly wysiwyg. There are most any lookup support for all of form edits and fee sheet integration for billing procedures with 837I pulling missing pieces in OpenEMR from form(need to document) i.e ub04 basically stands on it's own.
Once this is in codebase I have id'ed some bugs for others that need to be part of separate P.R for documentation.
Let me know what I need to do about sql/5_0_0-to-5_0_1_upgrade.sql conflict. Thought I fixed on this last commit!

bradymiller commented:
@sjpadgett 
don't worry about the sql/5_0_0-to-5_0_1_upgrade.sql conflict; this basically always happen whenever add entries to that script and not a big deal.
i'll plan to do a full review/test on this tonight and will hopefully get it into the codebase :)
-brady

bradymiller commented on composer.json:
```diff
@@ -76,11 +81,12 @@
             "OpenEMR\\Sample\\": "library/sample/src",
             "OpenEMR\\Reminder\\": "library/reminder/src",
             "OpenEMR\\Billing\\": "library/billing/src"
```
need a comma end the end of this line

bradymiller commented on sql/5_0_0-to-5_0_1_upgrade.sql:
```diff
@@ -510,4 +510,21 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
+#IfNotRow2D list_options list_id page_validation option_id add_edit_event#theform_prov
+INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `activity`) VALUES
+('page_validation', 'add_edit_event#theform_prov', '/interface/main/calendar/add_edit_event.php?prov=true', 170, '{}', 1);
+#EndIf
```
This is odd since this is already in the codebase (guessing is related to the merge conflict and I can easily deal with this when integrate it into codebase)

sjpadgett replied:
It was showing as part of conflict in git conflict resolution. I thought maybe was missed on prior commit after checked database.sql so I added from conflict resolution!

bradymiller commented:
overall looks good. let me see how it looks in terms of psr2. if not much work on that end, then I'll likely make the changes I mentioned in this review and get this into the codebase.

bradymiller commented:
Noted following bug in ub04 on billing page. If click the button to create the ubo4 (text or pdf), get the following error:
``````
The exit status code `126` says something went wrong: stderr: `sh: 1: /var/www/openemr/vendor/openemr/wkhtmltopdf-openemr/bin/linux/wkhtmltopdf64-linux: Permission denied ` stdout: `` command: /var/www/openemr/vendor/openemr/wkhtmltopdf-openemr/bin/linux/wkhtmltopdf64-linux --lowquality --grayscale --margin-bottom `0in` --margin-left `0.19in` --margin-right `0.19in` --margin-top `0.15in` --orientation `Portrait` --page-size `Letter` --no-outline --print-media-type --zoom `1.045` --keep-relative-links `/var/www/openemr/sites/default/documents/temp/knp_snappy59ad1a56459be3.05065270.html` `/var/www/openemr/sites/default/documents/temp/knp_snappy59ad1a5645b2f3.84233071.pdf`.
``````

bradymiller commented:
Question on what to do for the composer package. It looks like those binaries will take up 220MB (more than doubling OpenEMR's size :) )
Would it be better to make that a optional composer install (ie. user can do it after install if they need it for ub04)?

bradymiller commented:
To clarify, making composer optional would mean we would remove it from composer.json and then provide instructions how to install it(would be something like the command below):
composer require openemr/wkhtmltopdf-openemr

bradymiller commented:
quick note; this passed PSR2 testing :)

bradymiller commented:
Another PR here:
https://github.com/openemr/openemr/pull/1064
(I think this is ready to come in if ok with making openemr/wkhtmltopdf-openemr an optional module after install)

sjpadgett commented:
Yep, composer needs binaries removed. You take it out, I put it back, almost seems like I have a sub conscious mission here:) The permission error for binary is because pdf class does not have user write privileges on that directory to change binary file permission to exe. Same problem with nodejs forcing setup intervention at root. Suppose I need to come up with an install method for cases like this. I think Robert eluded to this with sub module suggestions. I'll also update repository readme with require command line for now and maybe start the node repository.
Once again Brady thanks, it is always fun and easy working with you.

sjpadgett commented:
@bradymiller  Should I do a commit to delete composer?

bradymiller commented:
@sjpadgett ,
Can we give the binaries execution privileges in the original composer repo?
And don't worry about the delete composer line. I'll do that (already did it in the above PR). Feel free to keep working here on anything else though before it gets into the codebase (i can easily migrate it over).
-brady

sjpadgett commented:
@bradymiller I had that thought a while back but got wayward. Not sure, but I am working that now along with running install scripts from composer and using bower/npm for nodejs. Need to get sorted.

sjpadgett commented:
@bradymiller git update-index --chmod=+x  wkhtmltopdf64-linux. Now how easy was that? Thanks for steering me back to the easy way 'cause as you know by now; The hard way is my way:)

bradymiller commented:
also, on the README, you can just remove the Compsoer.json Edit section since:
1. Am going to leave in the repository section in the composer.json script
2. Thus the user just needs to run the 'composer require...' statement to bring it in (and this also places an entry in the composer.json file)

sjpadgett commented:
Okay, looks like namespace has changed so having to update test server but I do know permissions did update correctly from composer.

bradymiller commented:
@sjpadgett ,
Just brought your code into the codebase! thanks!!
https://github.com/openemr/openemr/commit/e387e222fd9bca41ce27c4b93f3d0e4418e751cf

bradymiller commented:
@sjpadgett ,
Also just brought in a another commit into the codebase to sort out composer a bit(removed dev packages which was making installing packages too complicated), so now users can easily bring in your binaries now with the following command:
``````
composer require openemr/wkhtmltopdf-openemr
``````

bradymiller commented:
(btw, I just tested ub04 pdf creationg after doing above command and it worked like a charm!)

sjpadgett commented:
@bradymiller Terrific. This made for great test case that will carry forward with nodejs. Now with nodejs, angular and our own repositories part of OpenEMR, It'd be reasonable to expect to see more development using those resources.
